Vendor dependency cycle-free swarmkit

corhere opened the pull request:
Moby imports Swarmkit; Swarmkit no longer imports Moby. In order to accomplish this feat, Swarmkit has introduced a new plugin.Getter interface so it could stop importing our pkg/plugingetter package. This new interface is not entirely compatible with our
plugingetter.PluginGetter interface, necessitating a thin adapter.

Swarmkit had to jettison the CNM network allocator to stop having to import libnetwork as the cnmallocator package is deeply tied to libnetwork. Move the CNM network allocator into libnetwork, where it belongs. The package had a short an uninteresting Git history in the Swarmkit repository so no effort was made to retain history.

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
It must be placed inside the below triple backticks section:
-->
```markdown changelog

```

**- A picture of a cute animal (not mandatory but encouraged)**

corhere committed:
Vendor dependency cycle-free swarmkit

Moby imports Swarmkit; Swarmkit no longer imports Moby. In order to
accomplish this feat, Swarmkit has introduced a new plugin.Getter
interface so it could stop importing our pkg/plugingetter package. This
new interface is not entirely compatible with our
plugingetter.PluginGetter interface, necessitating a thin adapter.

Swarmkit had to jettison the CNM network allocator to stop having to
import libnetwork as the cnmallocator package is deeply tied to
libnetwork. Move the CNM network allocator into libnetwork, where it
belongs. The package had a short an uninteresting Git history in the
Swarmkit repository so no effort was made to retain history.

Signed-off-by: Cory Snider <csnider@mirantis.com>

thaJeztah commented on libnetwork/cnmallocator/provider_test.go:
```diff
@@ -0,0 +1,30 @@
+
+	"github.com/moby/swarmkit/v2/api"
+	"github.com/moby/swarmkit/v2/testutils"
+	"github.com/stretchr/testify/assert"
```
Looks like we pulled in testify here for the test (looking at the code, it's probably trivial to rewrite to gotest.tools or plain go?)

corhere replied:
Probably, [but that wouldn't get testify out of the dependency graph.](
https://github.com/moby/swarmkit/blob/df1e66110903b9bee3e377df5b790024268dc42d/manager/allocator/allocator_test_suite.go#L21-L25)
https://github.com/moby/moby/blob/6fe7928c8c9b29bcd2d0232c356e76134656901f/libnetwork/cnmallocator/allocator_test.go#L1-L11

If moby having a testify dependency is a problem I'll need to rewrite that test suite in Swarmkit.

corhere replied:
Discussed in maintainer call. Having multiple assertion libraries vendored makes it too easy to import the non-preferred library when writing new tests and have it slip past review. Rewriting these tests to use gotest.tools would not get rid of the testify dependency because of `allocator.RunAllocatorTests`; it would only make testify a transitive dependency. I will update this PR to add a lint rule to ban direct imports of testify packages.

corhere replied:
@thaJeztah you were right: [it was trivial to rewrite.](https://pkg.go.dev/gotest.tools/v3@v3.5.1/assert/cmd/gty-migrate-from-testify)

thaJeztah commented on libnetwork/cnmallocator/networkallocator_test.go:
```diff
@@ -0,0 +1,788 @@
+	"github.com/docker/docker/libnetwork/types"
+	"github.com/moby/swarmkit/v2/api"
+	"github.com/moby/swarmkit/v2/manager/allocator/networkallocator"
+	"github.com/stretchr/testify/assert"
```
Same here

corhere committed:
libn/cnmallocator: migrate tests to gotest.tools/v3

Apply command gotest.tools/v3/assert/cmd/gty-migrate-from-testify to the
cnmallocator package to be consistent with the assertion library used
elsewhere in moby.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
hack: block imports of vendored testify packages

While github.com/stretchr/testify is not used directly by any of the
repository code, it is a transitive dependency via Swarmkit and
therefore still easy to use without having to revendor. Add lint rules
to ban importing testify packages to make sure nobody does.

Signed-off-by: Cory Snider <csnider@mirantis.com>
